src/_langs/en/updates/2015-06-12-Media-Source-Extensions-for-Audio.markdown

```diff
@@ -0,0 +1,373 @@
+
+{% highlight javascript %}
+  // Xing padding is encoded as 24bits within the header.  Note: This code will
+  // only work for Layer3 Version 1 and Layer2 MP3 files with XING frame counts
+  // and gapless information.  See the following document for more details:
+  // [http://www.codeproject.com/Articles/8295/MPEG-Audio-Frame-Header](http://www.codeproject.com/Articles/8295/MPEG-Audio-Frame-Header)
+  var xingDataIndex = byteStr.indexOf('Xing');
+  if (xingDataIndex == -1) xingDataIndex = byteStr.indexOf('Info');
+  if (xingDataIndex != -1) {
+    // See section 2.3.1 in the link above for the specifics on parsing the Xing
+    // frame count.
+    var frameCountIndex = xingDataIndex + 8;
+    var frameCount = ReadInt(byteStr.substr(frameCountIndex, 4));
+
+    // For Layer3 Version 1 and Layer2 there are 1152 samples per frame.  See
+    // section 2.1.5 in the link above for more details.
+    var paddedSamples = frameCount * 1152;
+
+    // ... we'll cover this below.
+{% endhighlight %}
+
+Now that we have the total number of samples we can move on to reading out the number of padding samples. Depending on your encoder this may be written under a 'LAME' or 'Lavf' tag nested in the Xing header. Exactly 17 bytes after this header there are 3 bytes representing the front and end padding in 12-bits each respectively.
+
+{% highlight javascript %}
+    xingDataIndex = byteStr.indexOf('LAME');
+    if (xingDataIndex == -1) xingDataIndex = byteStr.indexOf('Lavf');
+    if (xingDataIndex != -1) {
+      // See [http://gabriel.mp3-tech.org/mp3infotag.html#delays](http://gabriel.mp3-tech.org/mp3infotag.html#delays) for details of
+      // how this information is encoded and parsed.
+      var gaplessDataIndex = xingDataIndex + 21;
+      var gaplessBits = ReadInt(byteStr.substr(gaplessDataIndex, 3));
+
+      // Upper 12 bits are the front padding, lower are the end padding.
+      frontPadding = gaplessBits >> 12;
+      endPadding = gaplessBits & 0xFFF;
+    }
+
+    realSamples = paddedSamples - (frontPadding + endPadding);
+  }
+
+  return {
+    audioDuration: realSamples * SECONDS_PER_SAMPLE,
+    frontPaddingDuration: frontPadding * SECONDS_PER_SAMPLE
+  };
+}
+{% endhighlight %}
+
+With that we have a complete function for parsing the vast majority of gapless content out there. Edge cases certainly abound though, so caution is recommended before using similar code in production.
+
+## Appendix C: On Garbage Collection
+
+Memory belonging to SourceBuffers is actively [garbage collected](http://en.wikipedia.org/wiki/Garbage_collection_(computer_science)) according to content type, platform specific limits, and the current play position. In Chrome, memory will first be reclaimed from already played buffers. However, if memory usage exceeds platform specific limits, it will remove memory from unplayed buffers.
```

Comment:
Thanks Joe!

On 19 June 2015 at 18:46, Joe Medley <notifications@github.com> wrote:

> In
> src/_langs/en/updates/2015-06-12-Media-Source-Extensions-for-Audio.markdown
> <https://github.com/google/WebFundamentals/pull/1381#discussion_r32851945>
> :
>
> > +
> > +    realSamples = paddedSamples - (frontPadding + endPadding);
> > +  }
> > +
> > +  return {
> > +    audioDuration: realSamples * SECONDS_PER_SAMPLE,
> > +    frontPaddingDuration: frontPadding * SECONDS_PER_SAMPLE
> > +  };
> > +}
> > +{% endhighlight %}
> > +
> > +With that we have a complete function for parsing the vast majority of gapless content out there. Edge cases certainly abound though, so caution is recommended before using similar code in production.
> > +
> > +## Appendix C: On Garbage Collection
> > +
> > +Memory belonging to SourceBuffers is actively [garbage collected](http://en.wikipedia.org/wiki/Garbage_collection_(computer_science)) according to content type, platform specific limits, and the current play position. In Chrome, memory will first be reclaimed from already played buffers. However, if memory usage exceeds platform specific limits, it will remove memory from unplayed buffers.
>
> SourceBuffers looks to be the name of an actual object or interface.
> Please add back ticks.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/WebFundamentals/pull/1381/files#r32851945>.
>